### PR TITLE
Add new routing to new content integration

### PR DESCRIPTION
**Pull request by alexander-schranz**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? |yes
| BC breaks? | yes
| Deprecations? | yes 
| Fixed tickets | fixes # 
| Related issues/PRs | https://github.com/sulu/sulu/pull/7726, https://github.com/sulu/skeleton/pull/276
| License | MIT
| Documentation PR | sulu/sulu-docs# 

#### What's in this PR?

Add new route implementation for the content bundle. This PR adds a few hacks and things we should revert before the stable release. Most things where added to keep compatibility to older bundles and routings.

#### Why?

Created the new route implementation in https://github.com/sulu/sulu/pull/7726

#### ToDo: Done in https://github.com/sulu/sulu/pull/7726

 - [x] Create Entity see https://github.com/sulu/sulu/pull/7726
 - [x] Childs are update see https://github.com/sulu/sulu/pull/7726
 - [x] Grand childs are update see https://github.com/sulu/sulu/pull/7726
 - [x] Multi Localization support see https://github.com/sulu/sulu/pull/7726
 - [x] Multi Site support see https://github.com/sulu/sulu/pull/7726
 - [x] Support for non Site connected routes (article)
    - [x] Site A with `/test` and Site B `/test` and article parent Site A `/test/article` 
 
#### Related Pull Requests

 - [x] History Handling https://github.com/sulu/sulu/pull/7830
 - [x] Add Multi Site tests https://github.com/sulu/sulu/pull/7831
 - [x] Add Multi Localization https://github.com/sulu/sulu/pull/7831
 - [x] Add CrossSite Site test (article (site: null) -> page (site: website)) https://github.com/sulu/sulu/pull/7834
 - [x] Add Test for url with same prefix are not effect on the same level https://github.com/sulu/sulu/pull/7836
 - [x] Tree Full Edit https://github.com/sulu/sulu/pull/7835
 - [x] Route for AutoIncrement Entities: https://github.com/sulu/sulu/pull/7859
 - [x] Dynamic Router https://github.com/sulu/sulu/pull/7860
 - [x] Locale Switcher https://github.com/sulu/sulu/pull/7874
 - [x] Implement in Content Bundle (THIS PR)

### CODE REVIEW

File: packages/content/config/services.xml [hunk 1]
```diff
@@ -101,7 +101,9 @@
             <argument type="service" id="sulu_content.content_structure_bridge_factory"/>
             <argument type="service" id="sulu_http_cache.cache_lifetime.resolver"/>
 
-            <tag name="sulu_route.defaults_provider"/>
+            <tag name="sulu_route.route_defaults_provider" resource-key="examples"/>
```
Comment (Prokyonn) [inline packages/content/config/services.xml:104]: I guess this is only a temporary solution?

File: packages/page/src/Infrastructure/Sulu/Route/WebspaceSiteRouteGenerator.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+            if (!\is_string($site)) {
+                throw new MissingRequestContextParameterException(RequestAttributeEnum::SITE->value);
+            }
+        }
+
+        $url = $this->webspaceManager->findUrlByResourceLocator($slug, null, $locale, $site, $requestContext->getHost(), $requestContext->getScheme());
+
+        if (null === $url) {
+            throw new \RuntimeException(\sprintf('No found for "%s" in locale "%s" and site "%s".', $slug, $locale, $site));
```
Comment (Prokyonn) [inline packages/page/src/Infrastructure/Sulu/Route/WebspaceSiteRouteGenerator.php]:
```suggestion
            throw new \RuntimeException(\sprintf('No url found for "%s" in locale "%s" and site "%s".', $slug, $locale, $site));
```

File: packages/route/src/Infrastructure/Doctrine/Repository/RouteRepository.php [hunk 3]
```diff
@@ -70,7 +70,7 @@ protected function createQueryBuilder(array $filters): QueryBuilder
         if (\array_key_exists('site', $filters)) {
             $site = $filters['site'] ?? null;
             $queryBuilder->andWhere(
-                null === $site ? 'route.site IS NULL' : 'route.site = :site'
+                null === $site ? 'route.site IS NULL' : '(route.site = :site OR route.site IS NULL)'
```
Comment (Prokyonn) [inline packages/route/src/Infrastructure/Doctrine/Repository/RouteRepository.php:73]: What is the use case for `site = null OR site = :site` ?

File: packages/content/config/services.xml (hunk 1 above)
Comment (alexander-schranz) [reply packages/content/config/services.xml:104]: @Prokyonn yes, we need kind of find out which `doctrine` entities exist at compile / compiler pass where doctrine service not exists yet. I couldn't find a quick way for this and decided for the alphe with this workaround.

File: packages/route/src/Infrastructure/Doctrine/Repository/RouteRepository.php (hunk 3 above)
Comment (alexander-schranz) [reply packages/route/src/Infrastructure/Doctrine/Repository/RouteRepository.php:73]: article bundle. The request analyzer will tell us we are on site `sulu` when opening the page `https://sulu.io/blog/some-blog` but we should still get the article which is not site specific back.

Comment (alexander-schranz) [reply packages/route/src/Infrastructure/Doctrine/Repository/RouteRepository.php:73]: @Prokyonn if its unclear we could go or transparency we can also go with `'siteOrNull' => $site`

Comment (Prokyonn) [reply packages/route/src/Infrastructure/Doctrine/Repository/RouteRepository.php:73]: Ahh makes sense 👍  no I don't think we need `siteOrNull` I just haven't thought about articles

**Review by Prokyonn — APPROVED** [review]

**Comment by alexander-schranz** [comment]
As there exists now both routing tables for the alpha part we need change the PHPCR MIgration bundle to use `ro_routes_next` table instead of `ro_routes` and migrate it to that. @Prokyonn
